listenbrainz/webserver/static/js/src/FollowButton.tsx

```diff
@@ -0,0 +1,137 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+import * as React from "react";
+import APIService from "./APIService";
+
+type FollowButtonProps = {
+  user: ListenBrainzUser;
+  loggedInUser: ListenBrainzUser;
+  loggedInUserFollowsUser: boolean;
+};
+
+type FollowButtonState = {
+  loggedInUserFollowsUser: boolean;
+  justFollowed: boolean;
+  hover: boolean;
+  error: boolean;
+};
+
+class FollowButton extends React.Component<
+  FollowButtonProps,
+  FollowButtonState
+> {
+  APIService: APIService;
+
+  constructor(props: FollowButtonProps) {
+    super(props);
+    this.state = {
+      loggedInUserFollowsUser: props.loggedInUserFollowsUser,
+      hover: false,
+      justFollowed: false,
+      error: false,
+    };
+
+    this.APIService = new APIService(`${window.location.origin}/1`);
+  }
+
+  toggleHover = () => {
+    const { hover } = this.state;
+    this.setState({ hover: !hover, justFollowed: false });
+  };
+
+  handleButtonClick = () => {
+    const { loggedInUserFollowsUser } = this.state;
+    if (loggedInUserFollowsUser) {
+      this.unfollowUser();
+    } else {
+      this.followUser();
+    }
+  };
+
+  followUser = () => {
+    const { user } = this.props;
+    this.APIService.followUser(user.name).then(({ status }) => {
+      if (status === 200) {
+        this.setState({ loggedInUserFollowsUser: true, justFollowed: true });
+      } else {
+        this.setState({ error: true });
+      }
+    });
+  };
+
+  unfollowUser = () => {
+    const { user } = this.props;
+    this.APIService.unfollowUser(user.name).then(({ status }) => {
+      if (status === 200) {
+        this.setState({ loggedInUserFollowsUser: false, justFollowed: false });
+      } else {
+        this.setState({ error: true });
+      }
+    });
+  };
+
+  getButtonDetails = (): { buttonClass: string; buttonText: string } => {
+    const { error, justFollowed, loggedInUserFollowsUser, hover } = this.state;
+
+    if (error) {
+      return {
+        buttonClass: "btn btn-sm btn-danger",
+        buttonText: "Something went wrong!",
+      };
+    }
+
+    if (justFollowed) {
+      return { buttonClass: "btn btn-sm btn-info", buttonText: "Following" };
+    }
+
+    if (loggedInUserFollowsUser) {
+      if (!hover) {
+        return { buttonClass: "btn btn-sm btn-info", buttonText: "Following" };
+      }
+      return { buttonClass: "btn btn-sm btn-danger", buttonText: "Unfollow" };
+    }
+
+    return { buttonClass: "btn btn-sm btn-warning", buttonText: "Follow" };
+  };
+
+  render() {
+    return (
+      <>
+        <div
+          id="follow-button"
+          onClick={this.handleButtonClick}
+          onKeyPress={this.handleButtonClick}
+          onMouseEnter={this.toggleHover}
```

Comment:
It would be better to pass the hover state in the callback rather than inverting the current state. That way the colors won't end up getting inverted because of some bug in browser code.

Reply:
Done.